### ISSUE TITLE
Improved security of prime selection?

### ISSUE THREAD

**Issue by tfdahlin** [DaylightingSociety/Paillier#5]
Note: Implementing this could slow down performance significantly.

I still need to do a lot of reading about the number theory behind this, but poor prime selection in RSA can lead to strong attacks against the system. For Diffie-Hellman keys, safe primes are used to prevent specific types of mathematical attacks. Safe primes are defined such that p = 2q + 1, where q is also prime. This prevents [small order subgroups of n](https://www.johndcook.com/blog/2017/01/12/safe-primes-sylow-theorems-and-cryptography/), which allow discrete logarithms to be computed more easily.

I'm unsure whether it would be beneficial to further attempt to find [strong cryptographic primes](https://en.wikipedia.org/wiki/Strong_prime#Definition_in_cryptography), so this requires more research.

If implemented, I believe that these improvements should probably be optional arguments when generating primes, depending on the significance of the performance impact.

**Comment by milo-trujillo** [comment]
Absolutely worth looking in to further. If this vulnerability applies to Paillier as well as it does in RSA and Diffie-Hellman, then I think we should make strong primes mandatory. Adding an optional argument for "enable high-security mode" or "lower security for performance" means exposing this tradeoff to users. Most of our end users presumably have _not_ read the relevant crypto papers to fully understand the implications of this choice and how it fits their threat-model. I think we should protect programmers from shooting themselves in the foot in this scenario.

**Comment by tfdahlin** [comment]
Definitely. If this applies to Paillier (I'm not even sure if there will be any papers that say one way or another, since y'know, it's Paillier; which means I might have to do a *really* deep dive into number theory), then we should enable the high-security mode by default and allow users to explicitly opt into a low-security mode.

**Comment by milo-trujillo** [comment]
What's the advantage of having a "low-security mode"? Yes, it's faster, but if it's insecure, why even provide the option? Is this a vulnerability that can only be exploited with supercomputers? Is there a reasonable threat model where people would want to be vulnerable in exchange for performance?

**Comment by tfdahlin** [comment]
It might come down to just how significant the performance impact is, as well as the severity of the vulnerability. I don't (yet) know enough to say whether the vulnerability is exploitable in practice without immense resources, and I also don't know the extent of the performance impact we would face. 

We currently use Rabin-Miller primality testing on each prime which, in its simplest form, performs at O(k*log(log(log(n)))), where k is the number of rounds performed (we do 50). This is pretty efficient, but the changes I mentioned in [#3](https://github.com/DaylightingSociety/Paillier/issues/3) will make it slightly less efficient in exchange for improved accuracy. I'm unsure how significantly those changes will affect the efficiency of the algorithm, just that it will slow it down somewhat. 

By selecting 'safe' primes, we compound on poor efficiency by requiring significantly more tests to be run. Not only will we go through more numbers before finding one that works (since we're looking for special types of primes), but we also have to perform the test on pairs of numbers at a time, effectively doubling the amount of numbers we test in the first place. I don't know how to calculate the big-O notation for this due to how the algorithm short-circuits on discovering a number is composite.

If we further opt to use what are called "cryptographically strong primes", which have even more strict requirements than safe primes that are used in Diffie-Hellman (again, I'm unsure if this is even useful for our application), then that means we must test even more numbers, again compounding the computational complexity.

If this increased complexity compounds too much, we may reach a point where the library is rendered effectively unusable through a combination of:
a) eliminating too many primes in search of a safe or cryptographically strong prime, and;
b) taking too long to check against probable primes
